File: python/0217-contains-duplicate.py
"""
LeetCode 217: Contains Duplicate

Problem:
Given an integer array nums, return true if any value appears at least twice in the array,
and return false if every element is distinct.

Example:
nums = [1,2,3,1] => True
nums = [1,2,3,4] => False
nums = [99,99] => True

Time Complexity: O(n) for hash set, O(n log n) for sorting
Space Complexity: O(n) for hash set, O(1) for sorting (with modified input)
"""


# Approach 1: Hash Set
def containsDuplicate_HashSet(nums):
    """
    Keep track of seen numbers using a set
    If we encounter a number already in the set, return True
    Time: O(n), Space: O(n)
    """
    seen = set()
    for num in nums:
        if num in seen:
            return True
        seen.add(num)
    return False


# Approach 2: Sorting
def containsDuplicate_Sorting(nums):
    """
    Sort the array, then check if any adjacent elements are equal
    Time: O(n log n), Space: O(1) or O(n) depending on sorting implementation
    """
    nums.sort()
    for i in range(len(nums) - 1):
        if nums[i] == nums[i + 1]:
            return True
    return False


# Approach 3: Set Length Comparison
def containsDuplicate_SetLength(nums):
    """
    If the length of the set is less than the length of the array,
    there must be duplicates
    Time: O(n), Space: O(n)
    """
    return len(set(nums)) < len(nums)


# Approach 4: Hash Map with Counter
def containsDuplicate_Counter(nums):
    """
    Count occurrences of each number
    Time: O(n), Space: O(n)
    """
    from collections import Counter

    count = Counter(nums)
    return any(v > 1 for v in count.values())


# Approach 5: Early Exit with Set
def containsDuplicate_EarlyExit(nums):
    """
    Similar to hash set but with more explicit early exit
    Time: O(n), Space: O(n)
    """
    if len(nums) <= 1:
        return False

    seen = set()
    for num in nums:
        if num in seen:
            return True
        seen.add(num)

    return False


# Test Cases
def test_containsDuplicate():
    test_cases = [
        ([1, 2, 3, 1], True),
        ([1, 2, 3, 4], False),
        ([99, 99], True),
        ([], False),
        ([1], False),
        ([1, 1], True),
        ([1, 2, 3, 4, 5, 1], True),
        ([-1, -1, -1], True),
        ([0, 1, 2, 3, 4], False),
    ]

    for nums, expected in test_cases:
        # Make copies since sorting modifies the array
        nums_copy1 = nums.copy()
        nums_copy2 = nums.copy()

        assert (
            containsDuplicate_HashSet(nums) == expected
        ), f"HashSet failed for {nums}"
        assert (
            containsDuplicate_Sorting(nums_copy1) == expected
        ), f"Sorting failed for {nums}"
        assert (
            containsDuplicate_SetLength(nums) == expected
        ), f"SetLength failed for {nums}"
        assert (
            containsDuplicate_Counter(nums) == expected
        ), f"Counter failed for {nums}"
        assert (
            containsDuplicate_EarlyExit(nums) == expected
        ), f"EarlyExit failed for {nums}"

        print(f"nums={nums} => {expected} ✓")


if __name__ == "__main__":
    test_containsDuplicate()
    print("\nAll tests passed!")

    # Example usage
    print(f"\nContains Duplicate Example: {containsDuplicate_HashSet([1, 2, 3, 1])}")

File: java/0217-contains-duplicate.java
/**
 * LeetCode 217: Contains Duplicate
 *
 * Problem:
 * Given an integer array nums, return true if any value appears at least twice in the array,
 * and return false if every element is distinct.
 *
 * Example:
 * nums = [1,2,3,1] => true
 * nums = [1,2,3,4] => false
 * nums = [99,99] => true
 */

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class ContainsDuplicate {

    // Approach 1: Hash Set
    public static boolean containsDuplicate_HashSet(int[] nums) {
        /**
         * Keep track of seen numbers using a set
         * Time: O(n), Space: O(n)
         */
        Set<Integer> seen = new HashSet<>();
        for (int num : nums) {
            if (seen.contains(num)) {
                return true;
            }
            seen.add(num);
        }
        return false;
    }

    // Approach 2: Sorting
    public static boolean containsDuplicate_Sorting(int[] nums) {
        /**
         * Sort the array, then check if any adjacent elements are equal
         * Time: O(n log n), Space: O(1) or O(n)
         */
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 1; i++) {
            if (nums[i] == nums[i + 1]) {
                return true;
            }
        }
        return false;
    }

    // Approach 3: Set Length Comparison
    public static boolean containsDuplicate_SetLength(int[] nums) {
        /**
         * If set length < array length, there are duplicates
         * Time: O(n), Space: O(n)
         */
        Set<Integer> set = new HashSet<>();
        for (int num : nums) {
            set.add(num);
        }
        return set.size() < nums.length;
    }

    // Approach 4: Early Exit
    public static boolean containsDuplicate_EarlyExit(int[] nums) {
        /**
         * Similar to hash set with explicit early exit
         * Time: O(n), Space: O(n)
         */
        if (nums.length <= 1) return false;

        Set<Integer> seen = new HashSet<>();
        for (int num : nums) {
            if (seen.contains(num)) {
                return true;
            }
            seen.add(num);
        }
        return false;
    }

    // Test Cases
    public static void main(String[] args) {
        int[][] testCases = {
            {1, 2, 3, 1},
            {1, 2, 3, 4},
            {99, 99},
            {1},
            {1, 1},
            {1, 2, 3, 4, 5, 1},
            {-1, -1, -1},
            {0, 1, 2, 3, 4}
        };

        boolean[] expectedResults = {true, false, true, false, true, true, true, false};

        for (int i = 0; i < testCases.length; i++) {
            int[] nums = testCases[i];
            boolean expected = expectedResults[i];

            // Make copies since sorting modifies the array
            int[] copy1 = nums.clone();
            int[] copy2 = nums.clone();

            assert containsDuplicate_HashSet(nums) == expected : "HashSet failed";
            assert containsDuplicate_Sorting(copy1) == expected : "Sorting failed";
            assert containsDuplicate_SetLength(copy2) == expected : "SetLength failed";
            assert containsDuplicate_EarlyExit(nums) == expected : "EarlyExit failed";

            System.out.println("Test case " + i + ": " + expected + " ✓");
        }

        System.out.println("\nAll tests passed!");

        // Example usage
        int[] example = {1, 2, 3, 1};
        System.out.println("\nContains Duplicate Example: " + containsDuplicate_HashSet(example));
    }
}
